tests/interfaces/test_batch_autograd.py

```diff
@@ -300,7 +300,7 @@ def cost(x, cache):
 
         expected_runs_ideal = 1  # forward pass
         expected_runs_ideal += 2 * N  # Jacobian
-        expected_runs_ideal += 2 * N + 1  # Hessian diagonal
+        expected_runs_ideal += N + 1  # Hessian diagonal

```

Comment:
Curious what this is?

Reply:
When using an autodiff framework to compute second derivatives, the autodiff framework will always first perform:

- a forward pass execution of the cost function,
- a backwards pass (first derivatives)
- a second backwards pass (second derivatives).

This is because the autodiff framework always assumes backprop, and forward passes are always needed before gradients can be accumulated when using backprop.

So summing up the number of expected executions, assuming a two-term shift rule and N parameters:

- Forward pass: 1 execution
- Backwards pass: 2*N executions
- Second backwards pass: N+1 executions for the diagonal of the Hessian (as the second-order two-term shift rule on the diagonal is `[f(0) - f(x_n + pi)] / 2`). On the off-diagonal, the second-order shift rule will have 4 terms, but we only need to worry about the upper-triangular elements due to symmetry. So this results in `4 * N * (N-1) /2` executions.

Reply:
Thanks so much!